dht_net.go

```diff
@@ -109,40 +108,62 @@ func (dht *IpfsDHT) handleNewMessage(s inet.Stream) bool {
 
 // sendRequest sends out a request, but also makes sure to
 // measure the RTT for latency measurements.
-func (dht *IpfsDHT) sendRequest(ctx context.Context, p peer.ID, pmes *pb.Message) (*pb.Message, error) {
-
-	ms, err := dht.messageSenderForPeer(ctx, p)
+func (dht *IpfsDHT) sendRequest(ctx context.Context, p peer.ID, req *pb.Message) (*pb.Message, error) {
+	ps, err := dht.getStream(ctx, p)
 	if err != nil {
 		return nil, err
 	}
-
 	start := time.Now()
-
-	rpmes, err := ms.SendRequest(ctx, pmes)
+	replyChan, err := ps.request(ctx, req)
 	if err != nil {
+		ps.reset()
 		return nil, err
 	}
-
-	// update the peer (on valid msgs only)
-	dht.updateFromMessage(ctx, p, rpmes)
-
-	dht.peerstore.RecordLatency(p, time.Since(start))
-	logger.Event(ctx, "dhtReceivedMessage", dht.self, p, rpmes)
-	return rpmes, nil
+	onReply := func(reply *pb.Message) {
+		dht.streamPool.put(ps, p)
+		dht.updateFromMessage(ctx, p, reply)
+		dht.peerstore.RecordLatency(p, time.Since(start))
+	}
+	select {
+	case reply, ok := <-replyChan:
+		if !ok {
+			return nil, ps.err()
+		}
+		onReply(reply)
+		return reply, nil
+	case <-ctx.Done():
+		go func() {
+			if reply, ok := <-replyChan; ok {
+				onReply(reply)
+			}
+		}()
+		return nil, ctx.Err()
+	}
 }
 
 // sendMessage sends out a message
-func (dht *IpfsDHT) sendMessage(ctx context.Context, p peer.ID, pmes *pb.Message) error {
-	ms, err := dht.messageSenderForPeer(ctx, p)
+func (dht *IpfsDHT) sendMessage(ctx context.Context, p peer.ID, pmes *pb.Message) (err error) {
```

Comment:
please avoid using named return values, confuses the hell out of readers.

Reply:
Named return values are fine to declare vars, as long as the returns are explicit IMO. I use that pattern.

Reply:
Right, I meant using them with implicit returns.

Reply:
Yup I tripped over the same thing and pointed it out

Reply:
This is much cleaner now. Note that named returns are still useful in a lot of cases, such as how it's used now.